Pull Merkle verifier code into its own package

AlCutter opened the pull request:
This allows clients to pull in less of Trillian if they only need to verify Merkle artifacts.
Leaves some type aliases etc. in the old location so as not to break existing folks.

This change results in the following transient dep packages no longer being pulled in by log verification clients:
```
< database/sql
< database/sql/driver
< github.com/golang/mock/gomock
< github.com/google/trillian/crypto/keys
< github.com/google/trillian/crypto/keys/der
< github.com/google/trillian/merkle
< github.com/google/trillian/merkle/compact
< github.com/google/trillian/merkle/smt
< github.com/google/trillian/monitoring
< github.com/google/trillian/storage
< github.com/google/trillian/storage/storagepb
< github.com/google/trillian/util/clock
```

### Checklist

- [X] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [ ] I have updated [documentation](docs/) accordingly (including the [feature implementation matrix](docs/Feature_Implementation_Matrix.md)).

mhutchinson commented:
@AlCutter have you noticed the build is failing with `flag provided but not defined: -alsologtostderr`. That's odd.

I'll continue to review.

mhutchinson commented:
How committed are you to "no breaking changes" here? `InMemoryMerkleTree` and its builder have migrated without any forward references which could be an issue. I'm not clear on what this refactoring into a new package (`memory`) buys us. I'll take another look tomorrow and it might be worth chatting about it too.

pphaneuf commented:
Note that the current structure pulls test code into production binaries (which has bitten me in the past, on a few occasions), so I'm in favour of this, personally. There's some comment saying that `InMemoryMerkleTree` is not part of the Trillian API, considering the legacy status of this repo (from before modules and all that), this is on the better defined side of things, I would say?

AlCutter commented:
> How committed are you to "no breaking changes" here? `InMemoryMerkleTree` and its builder have migrated without any forward references which could be an issue. I'm not clear on what this refactoring into a new package (`memory`) buys us. I'll take another look tomorrow and it might be worth chatting about it too.

You actually caught me mid-work on this (I got interrupted by meetings) :) 
The refactor of the memory thing is to break a cyclic dep introduced by the `merkle/verifier` package (`merkle` -> `m/verifier` for the aliases, and `m/verifier` -> `merkle` for the in memory tree used by the test).

The in memory stuff does explicitly say it's not part of the API, but if we feel strongly about it I can add type aliases for that too.

AlCutter commented:
> There's some comment saying that `InMemoryMerkleTree` is not part of the Trillian API, considering the legacy status of this repo (from before modules and all that), this is on the better defined side of things, I would say?

I've added a commit to make that explicit/enforced by moving it into an `internal` directory.

mhutchinson reviewed (commented):
I think this looks pretty good. It's easy to make a mistake doing this by eye though. I'd really like https://github.com/golang/exp/tree/master/apidiff to be able to show the diff here. In general I'd love this to be set up for all commits to show a report.

I'll take a break from this now, and when you're happy with the final state of this PR let me know and I'll take another look to see for anything I might be missing now.

AlCutter commented:
For postetrity:
```bash
$ go-apidiff --repo-path=. master

github.com/google/trillian/merkle
  Incompatible changes:
  - InMemoryMerkleTree: removed
  - NewInMemoryMerkleTree: removed
  - TreeEntry: removed
  - TreeEntryDescriptor: removed
  - VerifyMapInclusionProof: removed

github.com/google/trillian/merkle/coniks
  Incompatible changes:
  - Default: changed from github.com/google/trillian/merkle/hashers.MapHasher to *Hasher
  - New: changed from func(crypto.Hash) github.com/google/trillian/merkle/hashers.MapHasher to func(crypto.Hash) *Hasher

github.com/google/trillian/merkle/hashers
  Incompatible changes:
  - NewLogHasher: removed
  - NewMapHasher: removed
  - RegisterLogHasher: removed
  - RegisterMapHasher: removed
```

AlCutter commented:
```bash
 go-apidiff --repo-path=. master

github.com/google/trillian/merkle
  Incompatible changes:
  - InMemoryMerkleTree: removed
  - NewInMemoryMerkleTree: removed
  - TreeEntry: removed
  - TreeEntryDescriptor: removed
  - VerifyMapInclusionProof: removed
```

pphaneuf commented:
This tool is amazing! +1 on getting this on every PR!

pavelkalinnikov commented on CHANGELOG.md:
```diff
@@ -5,6 +5,11 @@
+   * Moved verifiers from `merkle` into `merkle/{log,map}verifier`sub-pacakges,
+     reducing the amount of extra baggage inadvertently pulled in by clients.
+ * Moved some packages intended for internal-only use into `internal` packages:
+   * InMemoryMerkleTree (indended to only be used by Trillian tests)
```
Nice!

pavelkalinnikov commented on client/log_verifier.go:
```diff
@@ -48,7 +48,7 @@ func NewLogVerifier(hasher hashers.LogHasher, pubKey crypto.PublicKey, sigHash c
 		PubKey:  pubKey,
 		SigHash: sigHash,
-		v:       merkle.NewLogVerifier(hasher),
+		v:       logverifier.NewLogVerifier(hasher),
```
`logverifier.New` would look more Go-style. Note that you don't need to preserve the original name because you have an alias.

AlCutter replied:
done

pavelkalinnikov commented on go.sum:
```diff
@@ -84,6 +84,7 @@ github.com/apache/beam v2.24.0+incompatible h1:oATgc+AEkjL3zE1e0f/7u3u6eBEpeaiTn
 github.com/apache/beam v2.24.0+incompatible/go.mod h1:/8NX3Qi8vGstDLLaeaU7+lzVEu/ACaQhYjeefzQ0y1o=
 github.com/apache/beam v2.26.0+incompatible h1:BFyRqKyUCaZDWVXOXtEjUVeAngilWOZzMG3KWUXFlZU=
 github.com/apache/beam v2.26.0+incompatible/go.mod h1:/8NX3Qi8vGstDLLaeaU7+lzVEu/ACaQhYjeefzQ0y1o=
+github.com/apache/beam v2.27.0+incompatible h1:Ua4pP8fXzq6AlwcLbOyvtdx+6QA+GNMA11xAIDrm+eY=
```
Are these changes caused by this PR? Consider factoring out.

AlCutter replied:
done

pavelkalinnikov replied:
This was added again (I guess it happens automatically when your rebuild).

AlCutter replied:
redone :)

pavelkalinnikov commented on integration/log.go:
```diff
@@ -363,7 +364,7 @@ func readbackLogEntries(logID int64, client trillian.TrillianLogClient, params T
 }
 
-func checkLogRootHashMatches(tree *merkle.InMemoryMerkleTree, client trillian.TrillianLogClient, params TestParameters) error {
+func checkLogRootHashMatches(tree *memory.InMemoryMerkleTree, client trillian.TrillianLogClient, params TestParameters) error {
```
Since you are moving it anyway (breaking change), consider renaming the package/type to `inmemory.MerkleTree`.

AlCutter replied:
done

pavelkalinnikov commented on internal/merkle/memory/memory_merkle_tree.go:
```diff
@@ -12,10 +12,9 @@
-
-// This is a fairly direct port of the C++ Merkle Tree. It has the same API and should have
-// similar performance. It keeps all its data in RAM and is not part of the Trillian API.
+// Package memory provides a fairly direct port of the C++ Merkle Tree from the CT repo.
```
TODO: Describe the type, instead of referencing to C++ repo.

AlCutter replied:
done

pavelkalinnikov commented on internal/merkle/memory/memory_merkle_tree.go:
```diff
@@ -12,10 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package merkle
```
The name of this file can be shortened (no need in `memory_` prefix since it's now in the path).

AlCutter replied:
done

pavelkalinnikov commented on merkle/logverifier/log_verifier.go:
```diff
@@ -0,0 +1,196 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logverifier
```
I'm skipping this file because I assume it's a copy-paste of the old one. Please confirm it is.

AlCutter replied:
it is, yup

pavelkalinnikov commented on client/map_verifier.go:
```diff
@@ -81,7 +81,7 @@ func (m *MapVerifier) VerifyMapLeafInclusion(smr *trillian.SignedMapRoot, leafPr
 // VerifyMapLeafInclusionHash verifies a MapLeafInclusion object against a root hash.
 func (m *MapVerifier) VerifyMapLeafInclusionHash(rootHash []byte, leafProof *trillian.MapLeafInclusion) error {
-	return merkle.VerifyMapInclusionProof(m.MapID, leafProof.GetLeaf(), rootHash, leafProof.GetInclusion(), m.Hasher)
+	return mapverifier.VerifyMapInclusionProof(m.MapID, leafProof.GetLeaf(), rootHash, leafProof.GetInclusion(), m.Hasher)
```
Similarly, "Map" can be removed from "VerifyMapInclusionProof".

AlCutter replied:
done

AlCutter commented:
/gcbrun
